server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java

```diff
@@ -0,0 +1,244 @@
+             * Exception does not implement hash code so we will compute the hash code based on the key set and the
+             * messages.
+             */
+            return Objects.hash(getKeys(failureByNames), getExceptionMessages(failureByNames));
+        }
+
+        private static List<String> getExceptionMessages(final Map<String, Exception> map) {
+            return map.values().stream().map(Throwable::getMessage).sorted(String::compareTo).collect(Collectors.toList());
+        }
+
+        private static List<String> getKeys(final Map<String, Exception> map) {
+            return map.keySet().stream().sorted(String::compareTo).collect(Collectors.toList());
+        }
+    }
+
+    public static final class TransportAction extends TransportMasterNodeAction<Request, Response> {
+
+        private final Client client;
+
+        @Inject
+        public TransportAction(TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                               ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, Client client) {
+            super(NAME, transportService, clusterService, threadPool, actionFilters, Request::new, indexNameExpressionResolver);
+            this.client = client;
+        }
+
+        @Override
+        protected String executor() {
+            return ThreadPool.Names.SAME;
+        }
+
+        @Override
+        protected Response read(StreamInput in) throws IOException {
+            return new Response(in);
+        }
+
+        @Override
+        protected void masterOperation(Task task,
+                                       Request request,
+                                       ClusterState state,
+                                       ActionListener<Response> listener) {
+            autoCreate(request, listener, client);
+        }
+
+        @Override
+        protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+            return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE, request.names.toArray(new String[0]));
+        }
+    }
+
+    static void autoCreate(Request request, ActionListener<Response> listener, Client client) {
+        // For now always redirect to the auto create index action, because only indices get auto created.
+        final AtomicInteger counter = new AtomicInteger(request.getNames().size());
+        final Map<String, Exception> results = new HashMap<>();
+        for (String name : request.getNames()) {
+            CreateIndexRequest createIndexRequest = new CreateIndexRequest();
+            createIndexRequest.index(name);
+            createIndexRequest.cause(request.getCause());
+            createIndexRequest.masterNodeTimeout(request.masterNodeTimeout());
+            createIndexRequest.preferV2Templates(request.getPreferV2Templates());
+            client.execute(AutoCreateIndexAction.INSTANCE, createIndexRequest, ActionListener.wrap(
```

Comment:
By delegating to a separate transport action, we risk that the retry on no longer master happens inside that transport action, making any decision we make on cluster state here invalid (or at least potentially stale). I would much prefer to do the cluster state update here.

That way we could also create all the indices and streams in one go, reducing the number of cluster states published.

I think auto-create does not have to be specific to data stream or index. It goes together with the document level privileges which are also agnostic of whether the data ends up in a data stream or an index.

Reply:
> I think auto-create does not have to be specific to data stream or index. It goes together with the document level privileges which are also agnostic of whether the data ends up in a data stream or an index.

Yes and this is another reason to keep all the auto creation logic in a single action.